### DISCUSSION TITLE
Using cp.Variable with jax

### DISCUSSION THREAD

**Discussion by yanalish** [cvxpy/cvxpy#1636] [Q&A]
Hello, 

I am sorry to bother you, but I am writing to see if you have any advice. I am currently trying to use cvxpy in combination with jax. Mainly I am trying to pass a cp.Variable to evaluate the gradient of a function. However jax does not recognise the type of the variable and thus spits out an error. Would you happen to know how I can go around this issue? 

As my code is entirely build in jax I would like to keep using it , but if there is no fix for this, is there another autodiff library that you know that works with the cp. Variable type?

The exact error message I am receiving is :
Argument 'var1309[0:6, 0] + [[ 0.00415967]
 [-0.49043015]
 [-0.38972838]
 [ 0.0413044 ]
 [-0.94767774]
 [ 1.05103195]][0:3, 0]' of type <class 'cvxpy.atoms.affine.index.index'> is not a valid JAX type.

Thank you very much in advance!

Yana

**Comment by SteveDiamond** [comment]
Hi Yana! cvxpy is not really designed to work with autodifferentiation software like Jax. You might be interested in [cvxpylayers](https://github.com/cvxgrp/cvxpylayers) though, which is a bit of a different idea but does interface with Jax.

**Reply by yanalish** [reply]
Hello! Thank you for the incredibly fast response! I didn't know of the existance of cvxpylayers, thank you for sharing it!  It seems however cxpylayers is slightly beyond what I needed. I do not wish to do backward passes. I just want to have constraints for example 

constraints = [grad(f, argnums = 0)(x) >= 0 ]
constraints = +[cp. SOC(t, grad(f, argnums = 0)(x)) ]

where x is a cp. Variable, t is some vector and grad is the function provided by jax.  The problem is the type of x is not accepted by grad. I am not sure cvxpylayers will help with this.

**Reply by SteveDiamond** [reply]
Got it. You can only construct a cvxpy problem using the [cvxpy atoms](https://www.cvxpy.org/tutorial/functions/index.html). You can't use Jax or anything like that.

**Reply by yanalish** [reply]
Thanks anyways! Sorry for all the questions.

Yana
